enigma/jwt.go

```diff
@@ -0,0 +1,174 @@
+eBIPmwKBgEZxhqa0gVvHQG/7Od69KWj4eJP28kq13RhKay8JOoN0vPmspXJo1HY3
+CKuHRG+AP579dncdUnOMvfXOtkdM4vk0+hWASBQzM9xzVcztCa+koAugjVaLS9A+
+9uQoqEeVNTckxx0S2bYevRy7hGQmUJTyQm3j1zEUR5jpdbL83Fbq
+-----END RSA PRIVATE KEY-----
+`},
+	{"../fosite-example/cert/rs256-public.pem",
+		`-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA4f5wg5l2hKsTeNem/V41
+fGnJm6gOdrj8ym3rFkEU/wT8RDtnSgFEZOQpHEgQ7JL38xUfU0Y3g6aYw9QT0hJ7
+mCpz9Er5qLaMXJwZxzHzAahlfA0icqabvJOMvQtzD6uQv6wPEyZtDTWiQi9AXwBp
+HssPnpYGIn20ZZuNlX2BrClciHhCPUIIZOQn/MmqTD31jSyjoQoV7MhhMTATKJx2
+XrHhR+1DcKJzQBSTAGnpYVaqpsARap+nwRipr3nUTuxyGohBTSmjJ2usSeQXHI3b
+ODIRe1AuTyHceAbewn8b462yEWKARdpd9AjQW5SIVPfdsz5B6GlYQ5LdYKtznTuy
+7wIDAQAB
+-----END PUBLIC KEY-----
+`,
+	},
+}
+
+// JWTEnigma : Container for jwt
+type JWTEnigma struct {
+	PrivateKey []byte
+	PublicKey  []byte
+	Claims     jwthelper.ClaimsContext
+	Headers    map[string]interface{}
+}
+
+// LoadCertificate : Read certificate from specified file
+func LoadCertificate(path string) ([]byte, error) {
+	if path == "" {
+		return nil, errors.Errorf("No path specified")
+	}
+
+	var rdr io.Reader
+	if f, err := os.Open(path); err == nil {
+		rdr = f
+		defer f.Close()
+	} else {
+		return nil, err
+	}
+	return ioutil.ReadAll(rdr)
+}
+
+func merge(a, b map[string]interface{}) map[string]interface{} {
+	for k, w := range b {
+		if _, ok := a[k]; ok {
+			continue
+		}
+		a[k] = w
+	}
+	return a
+}
+
+// GenerateChallenge : Generates a new authorize code or returns an error. set secret
+func (j *JWTEnigma) GenerateChallenge(secret []byte) (*Challenge, error) {
+	//** secret is not used with JWT using RSA **
+
+	// As per RFC, no overrides of header "alg"!
+	if _, ok := j.Headers["alg"]; ok {
+		return nil, errors.New("You may not override the alg header key.")
+	}
+
+	// As per RFC, no overrides of header "typ"!
+	if _, ok := j.Headers["typ"]; ok {
+		return nil, errors.New("You may not override the typ header key.")
+	}
+
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Claims = j.Claims
+	token.Header = merge(token.Header, j.Headers)
+	rsaKey, err := jwt.ParseRSAPrivateKeyFromPEM(j.PrivateKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var sig, sstr string
+
+	if sstr, err = token.SigningString(); err != nil {
+		return nil, err
+	}
+
+	if sig, err = token.Method.Sign(sstr, rsaKey); err != nil {
+		return nil, err
+	}
+
+	return &Challenge{
+		Key:       sstr,
+		Signature: sig,
+	}, nil
+}
+
+// ValidateChallenge :  Returns nil, if the code argument is a valid authorize code
+// and the signature matches the key.
+func (j *JWTEnigma) ValidateChallenge(secret []byte, t *Challenge) (err error) {
+	if t.Key == "" || t.Signature == "" {
+		return errors.New("Key and signature must both be not empty")
+	}
+
+	// Parse the token.
+	token, err := jwt.Parse(t.String(), func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
```

Comment:
Yes! I'll fix it. Though, earliest tomorrow noon. I recently switched to a
new job so there's little room for spare time  ;)
Den 13 jan 2016 23:30 skrev "Aeneas" <notifications@github.com>:

> In enigma/jwt.go
> <https://github.com/ory-am/fosite/pull/18#discussion_r49661287>:
>
> > +	return &Challenge{
> > +		Key:       sstr,
> > +		Signature: sig,
> > +	}, nil
> > +}
> > +
> > +// ValidateChallenge :  Returns nil, if the code argument is a valid authorize code
> > +// and the signature matches the key.
> > +func (j *JWTEnigma) ValidateChallenge(secret []byte, t *Challenge) (err error) {
> > +	if t.Key == "" || t.Signature == "" {
> > +		return errors.New("Key and signature must both be not empty")
> > +	}
> > +
> > +	// Parse the token.
> > +	token, err := jwt.Parse(t.String(), func(token *jwt.Token) (interface{}, error) {
> > +		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
>
> do you think you could write a test for this? maybe there is already one
> in the jwt-go repository. This check explicitly prevents this
> vulnerability <https://github.com/dgrijalva/jwt-go/issues/59> so it would
> be very cool to have a working test for this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ory-am/fosite/pull/18/files#r49661287>.
>